### PR TITLE
[CI] fix: flake8 rule B020: Found for loop that reassigns the iterable it is iterating with each iterable value.

### LINKED ISSUE #48067: Fix flake8 rule B020
This is a subtask of https://github.com/ray-project/ray/issues/47991. See the parent issue for more information.

https://github.com/PyCQA/flake8-bugbear

### PR DESCRIPTION
**Pull request by win5923**
## Why are these changes needed?

While running the pre-commit hook of flake8, the following error occurs if Python version is 3.12. It's because the version of flake8 is too old.

image
rename the loop variable to avoid shadowing the iterable.

Also, after resolving the B020 rule violation in `file_based_datasource.py`, running `./script/format.sh` resulted in a B023 violation, so I fixed that as well.

image

version:
- python: 3.12.7
- flake8: 7.1.1 
- flake8-bugbear: 24.8.19

## Related issue number

Closes #48067

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/datasource/file_based_datasource.py [hunk 1]
```diff
@@ -215,7 +215,9 @@ def read_files(
 
                 with _open_file_with_retry(
                     read_path,
-                    lambda: open_input_source(fs, read_path, **open_stream_args),
+                    lambda read_path=read_path: open_input_source(
+                        fs, read_path, **open_stream_args
+                    ),
```
Comment (jjyao) [inline python/ray/data/datasource/file_based_datasource.py:220]: Why this change?

Comment (win5923) [reply python/ray/data/datasource/file_based_datasource.py:220]: fix B023 violation when  I've run scripts/format.sh

**Review by jjyao — APPROVED** [review]
Nice!

**Review by hongpeng-guo — APPROVED** [review]

**Review by bveeramani — APPROVED** [review]
